src-java/flowhs-topology/flowhs-storm-topology/src/main/java/org/openkilda/wfm/topology/flowhs/fsm/haflow/validation/HaFlowValidationFsm.java

```diff
@@ -0,0 +1,340 @@
+        }
+
+        switchIdSet.addAll(extractSwitchIdSet(haFlow));
+
+        awaitingRulesSize = switchIdSet.size();
+        log.debug("Send commands to get the rules on the switches");
+        switchIdSet.forEach(switchId ->
+                getCarrier().sendSpeakerRequest(flowId, new DumpRulesForFlowHsRequest(switchId)));
+
+        Set<SwitchId> meterSwitchIds = new HashSet<>(haFlow.getEndpointSwitchIds());
+        meterSwitchIds.add(haFlow.getForwardPath().getYPointSwitchId());
+        Optional.ofNullable(haFlow.getProtectedForwardPath()).ifPresent(e -> meterSwitchIds.add(e.getYPointSwitchId()));
+
+        log.debug("Send commands to get the meters on the termination switches");
+        awaitingMetersSize = meterSwitchIds.size();
+        meterSwitchIds.forEach(switchId -> {
+            getCarrier().sendSpeakerRequest(flowId, new DumpMetersForFlowHsRequest(switchId));
+        });
+
+        log.debug("Send commands to get the groups on the termination switches");
+        awaitingGroupsSize = 2;
+
+        getCarrier().sendSpeakerRequest(flowId,
+                new DumpGroupsForFlowHsRequest(haFlow.getForwardPath().getYPointSwitchId()));
+        Optional.ofNullable(haFlow.getProtectedForwardPath()).ifPresent(protectedFlowPath -> {
+            getCarrier().sendSpeakerRequest(flowId,
+                    new DumpGroupsForFlowHsRequest(protectedFlowPath.getYPointSwitchId()));
+        });
+
+
+    }
+
+    protected void receivedRules(State from, State to,
+                                 Event event, Object context) {
+        FlowDumpResponse flowDumpResponse = (FlowDumpResponse) context;
+        log.info("The Switch rules received for the switch {}",
+                flowDumpResponse.getFlowSpeakerData().stream().findFirst().map(e -> e.getSwitchId().toString())
+                        .orElse("No switch Id"));
+        receivedRules.add(flowDumpResponse);
+        awaitingRulesSize--;
+        checkOfCompleteDataCollection();
+    }
+
+    protected void receivedMeters(State from, State to,
+                                  Event event, Object context) {
+        MeterDumpResponse meterDumpResponse = (MeterDumpResponse) context;
+        log.info("The Switch meters received for the switch {}",
+                meterDumpResponse.getMeterSpeakerData().stream().findFirst().map(e -> e.getSwitchId().toString())
+                        .orElse("No switch Id"));
+        receivedMeters.add(meterDumpResponse);
+        awaitingMetersSize--;
+        checkOfCompleteDataCollection();
+    }
+
+    protected void receivedGroups(State from, State to,
+                                  Event event, Object context) {
+        GroupDumpResponse groupDumpResponse = (GroupDumpResponse) context;
+        log.info("The Switch groups received for the switch {}",
+                groupDumpResponse.getGroupSpeakerData().stream().findFirst().map(e -> e.getSwitchId().toString())
+                        .orElse("No switch Id"));
```

Comment:
You have switch id in GroupDumpResponse. You can use it.

Reply:
fixed